src/java.base/share/classes/java/lang/reflect/Proxy.java

```diff
@@ -495,41 +495,17 @@ private static final class ProxyBuilder {
         private static final ClassLoaderValue<Boolean> reverseProxyCache =
             new ClassLoaderValue<>();
 
-        private static Class<?> defineProxyClass(Module m, List<Class<?>> interfaces) {
-            String proxyPkg = null;     // package to define proxy class in
-            int accessFlags = Modifier.PUBLIC | Modifier.FINAL;
-            boolean nonExported = false;
+        private record ProxyClassContext(Module module, String pkg, int accessFlags) { }
 
-            /*
-             * Record the package of a non-public proxy interface so that the
-             * proxy class will be defined in the same package.  Verify that
-             * all non-public proxy interfaces are in the same package.
-             */
-            for (Class<?> intf : interfaces) {
-                int flags = intf.getModifiers();
-                if (!Modifier.isPublic(flags)) {
-                    accessFlags = Modifier.FINAL;  // non-public, final
-                    String pkg = intf.getPackageName();
-                    if (proxyPkg == null) {
-                        proxyPkg = pkg;
-                    } else if (!pkg.equals(proxyPkg)) {
-                        throw new IllegalArgumentException(
-                                "non-public interfaces from different packages");
-                    }
-                } else {
-                    if (!intf.getModule().isExported(intf.getPackageName())) {
-                        // module-private types
-                        nonExported = true;
-                    }
-                }
-            }
+        private static Class<?> defineProxyClass(ProxyClassContext context, List<Class<?>> interfaces) {
+            String proxyPkg = context.pkg;
+            int accessFlags = context.accessFlags;
+            Module m = context.module;
 
-            if (proxyPkg == null) {
-                // all proxy interfaces are public and exported
+            if (Modifier.isPublic(accessFlags)) {
+                // all proxy interfaces are public
                 if (!m.isNamed())
                     throw new InternalError("unnamed module: " + m);
-                proxyPkg = nonExported ? PROXY_PACKAGE_PREFIX + "." + m.getName()
-                                       : m.getName();
             } else if (proxyPkg.isEmpty() && m.isNamed()) {
```

Comment:
With the refactoring, it may be easier to understand to make this check for both public and non-public access (i.e. drop the `else`) for clarity - a named module can't have unnamed package.

Reply:
> a named module can't have unnamed package

Since this is mandated by the Java Language Specification 7.4.2, I am tempted to change the thrown exception to `InternalError`, but I cannot find any restriction in the JVM Specification that prevents declaring a package-private superinterface in the unnamed package of a named module. If we can find a reference in the JVM Specification, I'm more than glad to refer to that in a comment and throw `InternalError` instead.